### DISCUSSION TITLE
Modular phase separation principle

### DISCUSSION THREAD

**Discussion by Ericson2314** [ghc-proposals/ghc-proposals#450] [General]
#448 has a "lexical scoping principle". I would love it we would commit to a generalization of this where:

- Parsing doesn't depend on anything
- Renaming doesn't depend on type-checking of this module *or it's imports*, and not the body/unfolding/implementation of any definition/interface 
- Type checking doesn't depend on Desugaring of this module *or it's imports*, and not the body/unfolding/implementation of any definition/interface

and

- Desugaring doesn't depend on the Optimization of this module, and only *optionally* on it's imports, and not the body/unfolding/implementation of any definition/interface
  - The fundamental purpose of worker/wrappers is to enable as many optimizations while de-abstracting definitions as little as possible.
- Optimization doesn't depend on Codegen of this module or it's imports.
- Codegen doesn't depend on the execution of this module or it's imports.

https://gitlab.haskell.org/ghc/ghc/-/issues/14095 emphsizes the implementation-side gains of this sort of policy, but the top half I think is good as a pure language design principle independent of performance. It's for the pure language design reasons that I open this thread.

**Comment by int-index** [comment]
This sounds to be at odds with TH. I can call e.g. `makeLenses`, which will generate new identifiers, so the renaming of the rest of the module will depend on running this splice.

**Comment by Ericson2314** [comment]
@int-index Yes it is, but in sort of a Hegelian thesis-and-antithesis-waiting-for-synthesis way. Concretely, I think the chief purpose of https://github.com/ghc-proposals/ghc-proposals/pull/412 with quote imports is to minimize the violation as little as possible by introduction an *orthogonal* notion of phase separation, and then carefully reformulating all the various restrictions with the *two dimensional phases :scream:* and whatnot so e.g.
```haskell
a = ...
$(f [| a |]) -- naively a cycle between the main stage and the previous splice stage 
```
still works.
